contracts/hydro/src/testing_compounder_cap.rs

```diff
@@ -0,0 +1,555 @@
+    contract::{execute, instantiate, sudo},
+    lsm_integration::get_total_power_for_round,
+    msg::ExecuteMsg,
+    state::{EXTRA_LOCKED_TOKENS_CURRENT_USERS, EXTRA_LOCKED_TOKENS_ROUND_TOTAL, LOCKED_TOKENS},
+    testing::{
+        get_address_as_str, get_default_instantiate_msg, get_message_info, IBC_DENOM_1,
+        IBC_DENOM_2, IBC_DENOM_3, ONE_DAY_IN_NANO_SECONDS, VALIDATOR_1, VALIDATOR_1_LST_DENOM_1,
+        VALIDATOR_2, VALIDATOR_2_LST_DENOM_1, VALIDATOR_3, VALIDATOR_3_LST_DENOM_1,
+    },
+    testing_lsm_integration::set_validator_infos_for_round,
+    testing_mocks::{
+        custom_interchain_query_mock, denom_trace_grpc_query_mock, mock_dependencies, ICQMockData,
+        MockQuerier,
+    },
+    testing_validators_icqs::get_mock_validator,
+    validators_icqs::TOKENS_TO_SHARES_MULTIPLIER,
+};
+
+const ROUND_LENGTH: u64 = 30 * ONE_DAY_IN_NANO_SECONDS;
+const LOCK_EPOCH_LENGTH: u64 = ROUND_LENGTH;
+const TEN_DAYS_IN_NANOS: u64 = 10 * ONE_DAY_IN_NANO_SECONDS;
+const FIRST_ROUND_START: Timestamp = Timestamp::from_nanos(1737540000000000000); // Wednesday, January 22, 2025 10:00:00 AM
+const INITIAL_BLOCK_HEIGHT: u64 = 19_185_000;
+const BLOCKS_PER_DAY: u64 = 35_000;
+
+// 1.  Round 0: Have 3 users fill the total cap by locking 3 different tokens for different duration (1, 6, 12 rounds).
+// 2.  Round 0: Update config to increase total_cap and set known_users_cap starting from round 1.
+// 3.  Round 0: Update config to close the known_users_cap after some time in round 1.
+// 4.  Round 0: Update all validator power ratios to verify that the total voting power changes, and users
+//     voting power also gets updated proportinally.
+// 5.  Round 1: Have the first known user unlock the expired lockup, to test voting power computation for previous round.
+// 6.  Round 1: Have the first known user lock some tokens in public_cap, then a completely new user lock tokens
+//     in public cap (try more than allowed, then lock below public_cap).
+// 7.  Round 1: Have the known user from previous step lock more to fill the public_cap and some more into known_users_cap.
+// 8.  Round 1: Have the same known user try to lock in known_users_cap more than it should be allowed.
+// 9.  Round 1: Have the same known user lock the most it should be allowed in the known_users_cap.
+// 10. Round 1: Have other two known users lock as much as they should be allowed in the known_users_cap.
+// 11. Round 1: Update config to increase total_cap and set known_users_cap starting from round 2.
+// 12. Round 1: Update config to close the known_users_cap after some time in round 2.
+// 13. Round 2: Have a completely new user lock tokens to fill up the public_cap, then try to lock more.
+// 14. Round 2: Have a known user lock maximum allowed in known users cap.
+// 15. Round 2: Advance the chain to end the known_users_cap duration and have a user from step #13 lock
+//     additional amount that matches the entire amount previously reserved for known_users_cap.
```

Comment:
In principle, I think I would have preferred a more focussed table-based test here, e.g.:
Try to have a user lock up:
* the public cap is large enough -> works and their known user cap remains unfilled
* public cap is not large enough, but known user cap is -> works, and their known user cap is partially filled now
* total is not large enough -> fails
* total would be large enough, but users known user cap is too small -> fails

I think we can leave this as-is, but just in general I think the long tests that try to test many things in one test feel less maintainable (for sure this is not the only place in the codebase where we have tests like this though).